### PR TITLE
Refactoring and cleanup

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  | -

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,13 +6,112 @@
...
+$identityRepository = getIdentityRepository(); // \Yiisoft\Auth\IdentityRepositoryInterface
+$authenticationMethod = new \Yiisoft\Auth\Method\HttpBasic($identityRepository);
+
+$middleware = new \Yiisoft\Auth\Middleware\Authentication(
+    $authenticationMethod,
+    $responseFactory, // PSR-17 ResponseFactoryInterface
+    $failureHandler // optional, \Yiisoft\Auth\Handler\AuthenticationFailureHandler by default
+);
+
+$middlewareDiapatcher->addMiddleware($middleware);
```
Comment (rustamwin) [inline README.md]:
```suggestion
$middlewareDispatcher->addMiddleware($middleware);
```

File: src/Method/QueryParameter.php [hunk 2]
```diff
@@ -30,7 +26,7 @@ public function __construct(IdentityRepositoryInterface $identityRepository)
 
     public function authenticate(ServerRequestInterface $request): ?IdentityInterface
     {
-        $accessToken = $request->getQueryParams()[$this->tokenParam] ?? null;
+        $accessToken = $request->getQueryParams()[$this->parameterName] ?? null;
         if (is_string($accessToken)) {
             return $this->identityRepository->findIdentityByToken($accessToken, get_class($this));
```
Comment (rustamwin) [inline src/Method/QueryParameter.php]:
```suggestion
            return $this->identityRepository->findIdentityByToken($accessToken, self::class);
```

File: src/Middleware/Authentication.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        return $handler->handle($request);
+    }
+
+    /**
+     * @param array $optional Patterns to match to consider the given request URI path optional.
+     * @return self
+     */
+    public function withOptional(array $optional): self
```
Comment (rustamwin) [inline src/Middleware/Authentication.php]: How about adding a method that combines the previous `optional`?

**Review by rustamwin — CHANGES_REQUESTED** [review]

File: src/Method/Composite.php [hunk 4]
```diff
@@ -4,43 +4,37 @@
...
     {
-        foreach ($this->authMethods as $i => $auth) {
-            if (!$auth instanceof AuthInterface) {
-                $this->authMethods[$i] = $auth = $this->container->get($auth);
-                if (!$auth instanceof AuthInterface) {
-                    throw new \RuntimeException(get_class($auth) . ' must implement ' . AuthInterface::class);
-                }
+        foreach ($this->authenticationMethods as $authenticationMethod) {
+            if (!$authenticationMethod instanceof AuthenticationMethodInterface) {
+                throw new \RuntimeException('Authentication method must be an instance of ' . AuthenticationMethodInterface::class . '.');
```
Comment (romkatsu) [inline src/Method/Composite.php:34]: This case is not covered by tests

File: src/Method/HttpHeader.php [hunk 5]
```diff
@@ -41,7 +37,7 @@ public function __construct(IdentityRepositoryInterface $identityRepository)
 
     public function authenticate(ServerRequestInterface $request): ?IdentityInterface
     {
-        $authToken = $this->getAuthToken($request);
+        $authToken = $this->getAuthenticationToken($request);
         if ($authToken !== null) {
             return $this->identityRepository->findIdentityByToken($authToken, get_class($this));
```
Comment (romkatsu) [inline src/Method/HttpHeader.php]:
```suggestion
            return $this->identityRepository->findIdentityByToken($authToken, static::class);
```

**Review by romkatsu — APPROVED** [review]

File: src/Middleware/Authentication.php (hunk 3 above)
Comment (samdark) [reply src/Middleware/Authentication.php]: I don't see a use-case for it.

**Review by rustamwin — APPROVED** [review]

**Pull request merged**